cephfs: set PV/PVC and clusterName details on the subvolume as metadata

pkalever opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This helps Monitoring solutions without access to Kubernetes clusters to display the details of the PV/PVC/NameSpace in their dashboard.         

## Is there anything that requires special attention ##

~~Needs: https://github.com/ceph/go-ceph/pull/691
Just for the CI to pass I have included go-ceph custom branch and hence `DNM`, but the reviews can continue for other commits.~~


## Related issues ##

Fixes: #2875


---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

Commit:
cephfs: add set/Get/List/Remove metadata utility functions

Add utility functions to set/Get/List/Remove PV/PVC/PVCNamespace metadata
on subvolume.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
cephfs: set PV/PVC details on the subvolume as metadata on create

This helps Monitoring solutions without access to Kubernetes clusters to
display the details of the PV/PVC/NameSpace in their dashboard.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
cephfs: set metadata on restart of provisioner pod

Make sure to set metadata when subvolume exist, i.e. if the provisioner pod
is restarted while createVolume is in progress, say it created the subvolume
but didn't yet set the metadata.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
e2e: add test cases for subvolume metadata validation

create a PVC and check PVC/PV metadata on cephFS subvolume

Fixes: #2875
Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
@Mergifyio rebase

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -402,6 +406,80 @@ var _ = Describe(cephfsType, func() {
+				if err != nil {
+					e2elog.Failf("failed to create PVC: %v", err)
+				}
+
```
add validateSubvolumeCount and  validateOmapCount validations

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -402,6 +406,80 @@ var _ = Describe(cephfsType, func() {
+					e2elog.Failf("expected pvcName %q got %q", pvc.Name, pvcName)
+				}
+
+				pvcNamespace, stdErr, err := execCommandInToolBoxPod(f,
```
move this to a helper function as we need to do same validation for PVC restore and PVC clone tests

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,86 @@
+	return fsa.RemoveMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
+}
+
+// ForceRemoveMetadata attempt to forcefully remove custom metadata set
+// on the subvolume in a volume using the metadata key.
+func (s *subVolumeClient) ForceRemoveMetadata(key string) error {
+	fsa, err := s.conn.GetFSAdmin()
+	if err != nil {
+		return err
+	}
+
+	return fsa.ForceRemoveMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
+}
```
where this is getting called?

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,86 @@
+
+// ListMetadata lists custom metadata (key-value pairs) set on the subvolume
+// in a volume.
+func (s *subVolumeClient) ListMetadata() (map[string]string, error) {
```
same here as well? if this is not getting called please remove it.

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,86 @@
+func (s *subVolumeClient) UnsetAllMetadata(keys []string) error {
+	for _, key := range keys {
+		err := s.RemoveMetadata(key)
+		// TODO: replace string comparison with errno.
```
please open an issue to track the TODO

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -402,6 +406,80 @@ var _ = Describe(cephfsType, func() {
 				}
 			})
 
+			By("create a PVC and check PVC/PV metadata on CephFS subvolume", func() {
```
Add e2e for pvc restore and pvc clone.

pkalever replied:
This will be part of a different PR: https://github.com/ceph/ceph-csi/pull/3168

Madhu-1 reviewed (changes requested):
requesting changes to avoid accidental merge as go-ceph changes are not available in the go-ceph release yet.

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -402,6 +406,80 @@ var _ = Describe(cephfsType, func() {
+				}
+				err = deleteResource(cephFSExamplePath + "storageclass.yaml")
+				if err != nil {
+					e2elog.Failf("failed to delete CephFS storageclass: %v", err)
```
to test these changes in we also need a new ceph release and also we need to make sure we support older ceph clusters also. need extra validation to check if ceph cluster supports these API's or not.

Commit:
cephfs: set cluster Name as metadata on the subvolume

This change helps read the cluster name from the cmdline args,
the provisioner will set the same on the subvolume.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
cephfs: unset cluster Name metadata

unsets the cluster name metadata key and value on the subvolume

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
deploy: add cluster name in the templates

added in helm charts which should help users.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
doc: add documentation about --clustername option

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Commit:
e2e: add tests to validate cluster name

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Madhu-1 reviewed (commented):
We need to support older ceph versions that do not support metadata operations. if ceph doesnot support metadata operations we need to skip setting  metadata

pkalever commented:
> We need to support older ceph versions that do not support metadata operations. if ceph doesnot support metadata operations we need to skip setting metadata

Not just Quincy, the patches were backported to Pacific and Octopus I guess. I think ideally we should check on all those versions, right?

As we also have snapshot metadata-related changes in a different PR https://github.com/ceph/ceph-csi/pull/3168, is this something that we can cover in a separate PR?

pkalever commented on e2e/cephfs_helper.go:
```diff
@@ -212,6 +212,39 @@ func listCephFSSubVolumes(f *framework.Framework, filesystem, groupname string)
+		return metadata, fmt.Errorf("error listing subvolume metadata %v", stdErr)
+	}
+
+	err = json.Unmarshal([]byte(stdout), &metadata)
```
@Madhu-1 @nixpanic do you see anything wrong here, unmarshal is failing for some reason?

In the previous run, I had added a log to print stdout and this is how it looks:

```
Jul 18 14:48:08.414: INFO: ExecWithOptions: execute(POST https://192.168.39.245:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-6944779c47-z9q6b/exec?command=%2Fbin%2Fsh&command=-c&command=ceph+fs+subvolume+metadata+ls+myfs+csi-vol-3ca56d9c-06a0-11ed-85d9-aedfd69586e0+--group_name%3De2e+--format%3Djson&container=rook-ceph-tools&container=rook-ceph-tools&stderr=true&stdout=true)
Jul 18 14:48:10.685: INFO: stdout=
{
    "csi.ceph.com/cluster/name": "k8s-cluster-1",
    "csi.storage.k8s.io/pv/name": "pvc-8552822b-6172-447b-9065-5a0d5954929a",
    "csi.storage.k8s.io/pvc/name": "csi-cephfs-pvc",
    "csi.storage.k8s.io/pvc/namespace": "cephfs-8081"
}
 stdErr= err=json: cannot unmarshal object into Go value of type []e2e.cephfsSubvolumeMetadata
Jul 18 14:48:10.685: FAIL: failed to list subvolume metadata: json: cannot unmarshal object into Go value of type []e2e.cephfsSubvolumeMetadata
```

pkalever replied:
I think I got the problem

pkalever commented:
> > We need to support older ceph versions that do not support metadata operations. if ceph doesnot support metadata operations we need to skip setting metadata
> 
> Not just Quincy, the patches were backported to Pacific and Octopus I guess. I think ideally we should check on all those versions, right?

Maybe we can just look at the ceph CLI for available metadata options and decide whether the ceph version supports it or not? 

 
 As we also have snapshot metadata-related changes in a different PR #3168, is this something that we can cover in a separate PR?

Madhu-1 commented:
> > > We need to support older ceph versions that do not support metadata operations. if ceph doesnot support metadata operations we need to skip setting metadata
> > 
> > 
> > Not just Quincy, the patches were backported to Pacific and Octopus I guess. I think ideally we should check on all those versions, right?
> 
> Maybe we can just look at the ceph CLI for available metadata options and decide whether the ceph version supports it or not?
> 
> As we also have snapshot metadata-related changes in a different PR #3168, is this something that we can cover in a separate PR?

We need to check on the server side  same like https://github.com/ceph/ceph-csi/blob/devel/internal/cephfs/core/volume.go#L35-L40 and skip metadata if its not supported

Commit:
cephfs: handle metadata op-failures with unsupported ceph versions

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented:
> We need to check on the server side same like https://github.com/ceph/ceph-csi/blob/devel/internal/cephfs/core/volume.go#L35-L40 and skip metadata if its not supported

Thanks for the hint, that helped! Please take a look at the changes.

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

[Logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.21/runs/4589/nodes/107/log/?start=0)

pkalever commented:
/retest mini-e2e-helm_k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23
[Logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.23/runs/2445/nodes/107/log/?start=0)

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22
[Logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.22/runs/3612/nodes/89/log/?start=0)

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,177 @@
+	if _, keyPresent = clusterAdditionalInfo[s.clusterID]; !keyPresent {
+		clusterAdditionalInfo[s.clusterID] = &localClusterState{}
+	}
+	// resize subvolume when either it's supported, or when corresponding
+	// clusterID key was not present.
```
Need to update the comment.

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,177 @@
+	}
+	clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported
+
+	return "", fmt.Errorf("operation GetMetadata is not supported")
```
use errors.New

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,177 @@
+	}
+	// resize subvolume when either it's supported, or when corresponding
+	// clusterID key was not present.
+	if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unknown ||
+		clusterAdditionalInfo[s.clusterID].subVolMetadataState == supported {
+		fsa, err := s.conn.GetFSAdmin()
+		if err != nil {
+			return "", err
+		}
+
+		res, err := fsa.GetMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
+		if err == nil {
+			clusterAdditionalInfo[s.clusterID].subVolMetadataState = supported
+
+			return res, nil
+		}
+		var invalid fsAdmin.NotImplementedError
+		// In case the error is other than invalid command return error to the caller.
+		if !errors.As(err, &invalid) {
+			return "", err
+		}
+	}
+	clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported
+
+	return "", fmt.Errorf("operation GetMetadata is not supported")
```
```suggestion
notSupErr:=errors.New("operation GetMetadata is not supported")
if  clusterAdditionalInfo[s.clusterID].subVolMetadataState== unsupported {
	return "", notSupErr
}
fsa, err := s.conn.GetFSAdmin()
if err != nil {
	return "", err
}
res, err := fsa.GetMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
if err != nil {
  var invalid fsAdmin.NotImplementedError
 // In case the error is other than invalid command return error to the caller.
  if errors.As(err, &invalid) {
  	clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported
  	  err=notSupErr
  }
  return "",err
  }
clusterAdditionalInfo[s.clusterID].subVolMetadataState = supported
return res, nil
}
```

Above suggestions makes it more readable?

pkalever replied:
surely does.

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,177 @@
+// using the metadata key.
+func (s *subVolumeClient) RemoveMetadata(key string) error {
+	// keyPresent checks whether corresponding clusterID key is present in clusterAdditionalInfo
+	var keyPresent bool
+	// verify if corresponding clusterID key is present in the map,
+	// and if not, initialize with default values(false).
+	if _, keyPresent = clusterAdditionalInfo[s.clusterID]; !keyPresent {
+		clusterAdditionalInfo[s.clusterID] = &localClusterState{}
+	}
+	// resize subvolume when either it's supported, or when corresponding
+	// clusterID key was not present.
+	if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unknown ||
```
Move the logic of checking support to a helper function and reuse it?

pkalever replied:
No, your previous suggestion for code change is good.

nixpanic commented on internal/cephfs/core/metadata.go:
```diff
@@ -29,34 +32,82 @@ const (
 // name.
 func (s *subVolumeClient) GetMetadata(key string) (string, error) {
+	notSupErr := errors.New("operation GetMetadata is not supported")
+	if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unsupported {
```
this is a very long if-statement, and it repeats itself often. Much cleaner to have a helper function like
```go
func (s *subVolumeClient) supportsSubVolMetadata() bool {}
```

nixpanic commented on internal/cephfs/core/metadata.go:
```diff
@@ -29,34 +32,82 @@ const (
 	}
+	res, err := fsa.GetMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
+	if err != nil {
+		var invalid fsAdmin.NotImplementedError
```
this part should really not be copy/pasted. place is in the helper function mentioned above.

is this not simpler for checking the error type? No need to cast the error into `invalid` is it is not used anywhere.
```go
if errors.Is(err, fsAdmin.NotImplementedError) {
```

pkalever replied:
If we use it directly we see this compilation error:
```
# github.com/ceph/ceph-csi/internal/cephfs/core
internal/cephfs/core/metadata.go:35:42: fsAdmin.NotImplementedError (type) is not an expression
internal/cephfs/core/metadata.go:79:22: fsAdmin.NotImplementedError (type) is not an expression
internal/cephfs/core/metadata.go:105:21: fsAdmin.NotImplementedError (type) is not an expression
make: *** [Makefile:160: cephcsi] Error 2
```

nixpanic commented on internal/cephfs/core/metadata.go:
```diff
@@ -17,46 +17,97 @@ limitations under the License.
+	var invalid fsAdmin.NotImplementedError
+	if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unsupported {
+		return false
+	} else if err != nil && errors.Is(err, &invalid) {
```
just use `errors.Is(err, fsAdmin.NotImplementedError)`, no need to assign it to `invalid` if you don't use that variable.

pkalever replied:
```
[0] pkalever 😎 ceph-csi✨ make
cephcsi image settings: quay.io/cephcsi/cephcsi version canary
if [ ! -d ./vendor ]; then (go mod tidy && go mod vendor); fi
GOOS=linux go build -tags=quincy,ceph_preview,ceph_ci_untested -mod vendor -a -ldflags '-X github.com/ceph/ceph-csi/internal/util.GitCommit=310cbc2025acc7a7703ce67582f1060c13f3d735 -X github.com/ceph/ceph-csi/internal/util.DriverVersion=canary' -o _output/cephcsi ./cmd/
# github.com/ceph/ceph-csi/internal/cephfs/core
internal/cephfs/core/metadata.go:35:41: fsAdmin.NotImplementedError (type) is not an expression
make: *** [Makefile:160: cephcsi] Error 2
[2] pkalever 👿 ceph-csi✨ git diff
diff --git a/internal/cephfs/core/metadata.go b/internal/cephfs/core/metadata.go
index 2dfdcf786..33be13419 100644
--- a/internal/cephfs/core/metadata.go
+++ b/internal/cephfs/core/metadata.go
@@ -30,10 +30,9 @@ const (
 )
 
 func (s *subVolumeClient) supportsSubVolMetadata(err error) bool {
-       var invalid fsAdmin.NotImplementedError
        if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unsupported {
                return false
-       } else if err != nil && errors.Is(err, &invalid) {
+       } else if err != nil && errors.Is(err, fsAdmin.NotImplementedError) {
                // In case the error is other than invalid command return error to the caller.
                clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported

```

nixpanic commented on internal/cephfs/core/metadata.go:
```diff
@@ -17,46 +17,97 @@ limitations under the License.
 // GetMetadata gets custom metadata on the subvolume in a volume based on key
 // name.
 func (s *subVolumeClient) GetMetadata(key string) (string, error) {
+	notSupErr := errors.New("operation GetMetadata is not supported")
```
`notSupErr` can be a package wide variable, call it `ErrSubVolMetadataNotSupported` so that others can use `errors.Is()` to handle it.

reduce this to
```
if !s.supportsSubVolMetadata(nil) {
    return "", ErrSubVolMetadataNotSupported
}
```

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,150 @@
+// GetMetadata gets custom metadata on the subvolume in a volume based on key
+// name.
+func (s *subVolumeClient) GetMetadata(key string) (string, error) {
+	if !s.supportsSubVolMetadata(nil) {
```
I dont think it's a better one. passing nil and err and making the decision on that one is not required; if this is required, have a new helper function to check the clusterAdditionalInfo and return the required error message.

pkalever replied:
I just optimized it a bit, may be passing `nil` directly made you think something is not right, will be okay if I make it to:

```
func (s *subVolumeClient) supportsSubVolMetadata(err error) bool {
        var invalid fsAdmin.NotImplementedError
        if clusterAdditionalInfo[s.clusterID].subVolMetadataState == unsupported {
                return false
        } else if err != nil && errors.Is(err, &invalid) {
                // In case the error is other than invalid command return error to the caller.
                clusterAdditionalInfo[s.clusterID].subVolMetadataState = unsupported

                return false
        }
        clusterAdditionalInfo[s.clusterID].subVolMetadataState = supported

        return true
}

// GetMetadata gets custom metadata on the subvolume in a volume based on key
// name.
func (s *subVolumeClient) GetMetadata(key string) (string, error) {
        var err error
        if !s.supportsSubVolMetadata(err) {
                return "", ErrSubVolMetadataNotSupported
        }
        fsa, err := s.conn.GetFSAdmin()
        if err != nil {
                return "", err
        }
        res, err := fsa.GetMetadata(s.FsName, s.SubvolumeGroup, s.VolID, key)
        if !s.supportsSubVolMetadata(err) {
                return "", ErrSubVolMetadataNotSupported
        }
        if err != nil {
                return "", err
        }

        return res, nil
}
```

pkalever replied:
Logically there is nothing wrong here!!

nixpanic replied:
I think I prefer two separate helpers to make their functioning clearer and easier to understand.

`s.isUnsupportedSubVolMetadata(err)`, pass the error, set either `supported` or `unsupported` and return 

`s.supportsSubVolMetadata()`, no error, just returns a bool. If the support was not checked before, it can do the checking if needed, or return `true` when in doubt. If it does checking, it should use `s.isUnsupportedSubVolMetadata(err)` to cache the support.

Madhu-1 reviewed (commented):
changes LGTM  but i have a concern on the way we are handing clusterAdditionalInfo check

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -311,6 +315,15 @@ func (cs *ControllerServer) CreateVolume(
 
+		if !volOptions.BackingSnapshot {
+			// Set metadata on restart of provisioner pod when subvolume exist
+			metadata := k8s.GetVolumeMetadata(req.GetParameters())
```
we can move this outside of if block and use it, For now we are calling `k8s.GetVolumeMetadata(req.GetParameters())` twice in createVolume

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -658,7 +679,7 @@ func (cs *ControllerServer) ControllerExpandVolume(
 	RoundOffSize := util.RoundOffCephFSVolSize(req.GetCapacityRange().GetRequiredBytes())
 
-	volClient := core.NewSubVolume(volOptions.GetConnection(), &volOptions.SubVolume, volOptions.ClusterID)
+	volClient := core.NewSubVolume(volOptions.GetConnection(), &volOptions.SubVolume, volOptions.ClusterID, "")
```
why not pass clusterName instead of an empty string in all cases, as we will have clusterName?

pkalever replied:
Sure I can pass wherever available.

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,155 @@
+)
+
+const (
+	// clusterNameKey cluster Key, set on RBD image.
```
this is for the cephfs subvolume, not RBD image?  reuse the clusterNameKey we have for RBD instead of defining separate one for cephfs?

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,155 @@
+	if !s.isUnsupportedSubVolMetadata(err) {
+		return ErrSubVolMetadataNotSupported
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
```
just return err should be enough?

Madhu-1 commented on internal/cephfs/core/metadata.go:
```diff
@@ -0,0 +1,155 @@
+	if !s.isUnsupportedSubVolMetadata(err) {
+		return ErrSubVolMetadataNotSupported
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
```
just return err?

Madhu-1 reviewed (commented):
some nits

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.24

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
@Mergifyio rebase

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

[Logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/upgrade-tests-cephfs/runs/6238/nodes/93/log/?start=0)

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
@Mergifyio rebase

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.24

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.24

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.22

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22
